Mathlib/RingTheory/Coalgebra.lean

```diff
@@ -0,0 +1,125 @@
+/-
+Copyright (c) 2023 Ali Ramsey. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ali Ramsey
+-/
+import Mathlib.RingTheory.TensorProduct
+
+/-!
+# Coalgebras
+
+In this file we define `Coalgebra`, and provide instances for:
+
+* Commutative rings: `CommRing.toCoalgebra`
+* Finitely supported functions: `Finsupp.instCoalgebra`
+
+## References
+
+* <https://en.wikipedia.org/wiki/Coalgebra>
+-/
+
+suppress_compilation
+
+universe u v
+
+open scoped TensorProduct
+
+/-- A coalgebra over a commutative ring `R` is a module over `R` equipped with a coassociative
+comultiplication `Δ` and a counit `ε` obeying the left and right conunitality laws. -/
+class Coalgebra (R : Type u) (A : Type v) [CommRing R] [AddCommGroup A] [Module R A] where
+  /-- The comultiplication of the coalgebra -/
+  comul : A →ₗ[R] A ⊗[R] A
+  /-- The counit of the coalgebra -/
+  counit : A →ₗ[R] R
+  /-- The comultiplication is coassociative -/
+  coassoc : TensorProduct.assoc R A A A ∘ₗ comul.rTensor A ∘ₗ comul = comul.lTensor A ∘ₗ comul
+  /-- The counit satisfies the left counitality law -/
+  rTensor_counit_comp_comul : counit.rTensor A ∘ₗ comul = TensorProduct.mk R _ _ 1
+  /-- The counit satisfies the right counitality law -/
+  lTensor_counit_comp_comul : counit.lTensor A ∘ₗ comul = (TensorProduct.mk R _ _).flip 1
+
+namespace Coalgebra
+
+section CommRingAddCommGroup
+
+variable {R : Type u} {A : Type v}
+variable [CommRing R] [AddCommGroup A] [Module R A] [Coalgebra R A]
+
+@[simp]
+theorem coassoc_apply (a : A) :
+    TensorProduct.assoc R A A A (comul.rTensor A (comul a)) = comul.lTensor A (comul a) :=
+  LinearMap.congr_fun coassoc a
+
+@[simp]
+theorem coassoc_symm_apply (a : A) :
+    (TensorProduct.assoc R A A A).symm (comul.lTensor A (comul a)) = comul.rTensor A (comul a) := by
+  rw [(TensorProduct.assoc R A A A).symm_apply_eq, coassoc_apply a]
+
```

Comment:
Can you also add the theorem that `(TensorProduct.assoc R A A A).symm ∘ₗ comul.lTensor A ∘ₗ comul = comul.rTensor A ∘ₗ comul`? It should follow from the lemma you just proved.